appengine/ndb/queries/snippets.py

```diff
@@ -0,0 +1,234 @@
+    return query
+
+
+def query_purchase_with_customer_key():
+    # [START purchase_with_customer_key_models]
+    class Customer(ndb.Model):
+        name = ndb.StringProperty()
+
+    class Purchase(ndb.Model):
+        customer = ndb.KeyProperty(kind=Customer)
+        price = ndb.IntegerProperty()
+    # [END purchase_with_customer_key_models]
+
+    def query_purchases_for_customer_via_key(customer_entity):
+        purchases = Purchase.query(
+            Purchase.customer == customer_entity.key).fetch()
+        return purchases
+
+    return Customer, Purchase, query_purchases_for_customer_via_key
+
+
+def query_purchase_with_ancestor_key():
+    # [START purchase_with_ancestor_key_models]
+    class Customer(ndb.Model):
+        name = ndb.StringProperty()
+
+    class Purchase(ndb.Model):
+        price = ndb.IntegerProperty()
+    # [END purchase_with_ancestor_key_models]
+
+    def create_purchase_for_customer_with_ancestor(customer_entity):
+        purchase = Purchase(parent=customer_entity.key)
+        return purchase
+
+    def query_for_purchases_of_customer_with_ancestor(customer_entity):
+        purchases = Purchase.query(ancestor=customer_entity.key).fetch()
+        return purchases
+
+    return (Customer, Purchase,
+            create_purchase_for_customer_with_ancestor,
+            query_for_purchases_of_customer_with_ancestor)
+
+
+def print_query():
+    print(Employee.query())
+    # -> Query(kind='Employee')
+    print(Employee.query(ancestor=ndb.Key(Manager, 1)))
+    # -> Query(kind='Employee', ancestor=Key('Manager', 1))
+
+
+def query_contact_with_city():
+    query = Contact.query(Contact.addresses.city == 'Amsterdam')
+    return query
+
+
+def query_contact_sub_entities_beware():
+    query = Contact.query(Contact.addresses.city == 'Amsterdam',  # Beware!
+                          Contact.addresses.street == 'Spear St')
+    return query
+
+
+def query_contact_multiple_values_in_single_sub_entity():
+    query = Contact.query(Contact.addresses == Address(city='San Francisco',
+                                                       street='Spear St'))
+    return query
+
+
+def query_properties_named_by_string_on_expando():
+    property_to_query = 'location'
+    query = FlexEmployee.query(ndb.GenericProperty(property_to_query) == 'SF')
+    return query
+
+
+def query_properties_named_by_string_for_defined_properties(keyword, value):
+    query = Article.query(Article._properties[keyword] == value)
+    return query
+
+
+def query_properties_named_by_string_using_getattr(keyword, value):
+    query = Article.query(getattr(Article, keyword) == value)
+    return query
+
+
+def order_query_results_by_property(keyword):
+    expando_query = FlexEmployee.query().order(ndb.GenericProperty('location'))
+
+    property_query = Article.query().order(Article._properties[keyword])
+
+    return expando_query, property_query
+
+
+def print_query_keys(query):
```

Comment:
It's included (like most of the stuff in this file) using an indented_block instead of a region_tag